File: js/npc.js
/**
 * NPC Class
 * Simple interactive characters in the game
 */

class NPC {
    constructor(x, y, npcIndex = 0, scale = 4, worldWidth = 1536, screenHeight = 1024, yOffset = 0) {
        this.x = x;
        this.y = y;
        this.npcIndex = npcIndex;
        this.scale = scale;
        this.worldWidth = worldWidth;
        this.screenHeight = screenHeight;
        this.yOffset = yOffset; // For fine-tuning foot alignment

        // Dimensions (371 pixels wide spritesheet containing 8 NPCs)
        this.baseWidth = 371 / 8; // 46.375px
        this.baseHeight = 74;
        this.width = 46 * scale;
        this.height = this.baseHeight * scale;

        // State
        this.state = 'walk'; // NPCs walk more by default
        this.facingLeft = Math.random() < 0.5;
        this.speed = 0.8 + Math.random() * 0.8; // Relaxed walking speed
        this.alive = true;
        this.leader = null;
        this.followDistance = 50 + Math.random() * 30;
        this.moveOnlyLeft = false; // For conveyor-belt like movement
        this.wrapAround = false;   // Respawn on the other side

        // Animation
        this.frameIndex = Math.random() * 4; // Start at random frame to prevent sync
        this.animationSpeed = 0.12;
        this.animations = {
            idle: [],
            walk: []
        };

        // Personality (AI variation)
        this.laziness = 0.002 + Math.random() * 0.01;      // Chance to stop walking (0.2% - 1.2%)
        this.restlessness = 0.005 + Math.random() * 0.02;  // Chance to start walking (0.5% - 2.5%)

        this.loaded = false;
        this.loadTextures();
    }

    async loadTextures() {
        try {
            const idleSheet = await Utils.loadImage('npc_textures/npc_idle.png');
            const walk0 = await Utils.loadImage('npc_textures/npc_walk0.png');
            const walk1 = await Utils.loadImage('npc_textures/npc_walk1.png');
            const walk2 = await Utils.loadImage('npc_textures/npc_walk2.png');

            if (idleSheet && walk0 && walk1 && walk2) {
                const startX = Math.round(this.npcIndex * this.baseWidth);
                const frameWidth = Math.ceil(this.baseWidth);

                this.animations.idle.push(this.extractFrame(idleSheet, startX, 0, frameWidth, this.baseHeight));
                this.animations.walk.push(this.extractFrame(walk0, startX, 0, frameWidth, this.baseHeight));
                this.animations.walk.push(this.extractFrame(walk1, startX, 0, frameWidth, this.baseHeight));
                this.animations.walk.push(this.extractFrame(walk2, startX, 0, frameWidth, this.baseHeight));
                this.animations.walk.push(this.extractFrame(walk1, startX, 0, frameWidth, this.baseHeight)); // Loop back through neutral

                this.loaded = true;
            } else {
                this.createFallbacks();
            }
        } catch (error) {
            console.error('Error loading NPC textures:', error);
            this.createFallbacks();
        }
    }

    extractFrame(sheet, x, y, width, height) {
        const canvas = document.createElement('canvas');
        canvas.width = width;
        canvas.height = height;
        const ctx = canvas.getContext('2d');
        ctx.imageSmoothingEnabled = false;
        ctx.drawImage(sheet, x, y, width, height, 0, 0, width, height);
        return canvas;
    }

    createFallbacks() {
        const colors = ['#32c832', '#3232c8', '#c8c832', '#c832c8', '#32c8c8'];
        const color = colors[this.npcIndex % colors.length];
        const fallback = Utils.createFallbackImage(this.baseWidth, this.baseHeight, color);
        this.animations.idle = [fallback];
        this.animations.walk = [fallback];
        this.loaded = true;
    }

    update(dt, npcs, player) {
        if (!this.loaded) return;

        // Player Avoidance (Personal Space)
        if (player && !this.leader && !this.id && !this.groupId) {
            const dx = this.x - player.x;
            const distance = Math.abs(dx);
            const dy = Math.abs((this.y + this.yOffset) - player.y);

            if (distance < 60 && dy < 50) {
                // If walking towards player, turn or stop
                const movingTowardsPlayer = (this.facingLeft && dx > 0) || (!this.facingLeft && dx < 0);
                if (movingTowardsPlayer && this.state === 'walk') {
                    if (Math.random() < 0.1) this.facingLeft = !this.facingLeft;
                    else if (Math.random() < 0.05) this.state = 'idle';
                }

                // Subtle push if very close
                if (distance < 40) {
                    this.x += (dx > 0 ? 1 : -1) * (dt / 16.67);
                }
            }
        }



        if (this.leader) {
            // Group following behavior
            const targetX = this.leader.x + (this.x < this.leader.x ? -this.followDistance : this.followDistance);
            const dx = targetX - this.x;

            if (Math.abs(dx) > 20) { // Slight buffer
                this.state = 'walk';
                this.facingLeft = dx < 0;
                this.x += (this.facingLeft ? -this.speed : this.speed) * (dt / 16.67);
            } else {
                // Mimic leader state when close
                this.state = this.leader.state;
                this.facingLeft = this.leader.facingLeft;
            }
        } else if (this.state === 'walk') {
            this.x += (this.facingLeft ? -this.speed : this.speed) * (dt / 16.67);

            // Check world boundaries
            if (this.wrapAround) {
                // Wrap around slightly off-screen for continuity
                if (this.x < -150) this.x = this.worldWidth + 150;
                else if (this.x > this.worldWidth + 150) this.x = -150;
            } else {
                if (this.x < 10) {
                    this.x = 11;
                    this.facingLeft = false;
                } else if (this.x > this.worldWidth - 10) {
                    this.x = this.worldWidth - 11;
                    this.facingLeft = true;
                }
            }

            // Rarely stop walking (Skip if moveOnlyLeft)
            if (!this.moveOnlyLeft && Math.random() < this.laziness) {
                this.state = 'idle';
            }
        } else {
            // Idle state: occasionally start walking
            // Skip random movement if NPC has an ID or Group (usually cinematic or following)
            if (!this.id && !this.groupId && Math.random() < this.restlessness) {
                this.state = 'walk';
                if (!this.moveOnlyLeft && Math.random() < 0.3) this.facingLeft = !this.facingLeft;
            }
        }

        // Update animation
        const anim = this.animations[this.state] || this.animations.idle;
        this.frameIndex += this.animationSpeed * (dt / 16.67);
        if (this.frameIndex >= anim.length) {
            this.frameIndex = 0;
        }
    }

    draw(ctx, dt = 16.67) {
        if (!this.loaded || !this.alive) return;

        const anim = this.animations[this.state] || this.animations.idle;
        const frame = anim[Math.floor(this.frameIndex)] || anim[0];

        if (!frame) return;

        ctx.save();
        ctx.imageSmoothingEnabled = false;

        // Foot alignment adjustment
        const drawY = this.y + this.yOffset;

        if (this.facingLeft) {
            ctx.scale(-1, 1);
            ctx.drawImage(frame, -this.x - this.width / 2, drawY - this.height, this.width, this.height);
        } else {
            ctx.drawImage(frame, this.x - this.width / 2, drawY - this.height, this.width, this.height);
        }

        ctx.restore();
    }

    getRect() {
        return {
            x: this.x - this.width / 2,
            y: this.y - this.height,
            width: this.width,
            height: this.height
        };
    }
}

class TeacherNPC extends NPC {
    constructor(x, y, name, scale = 4, worldWidth = 1536, screenHeight = 1024, yOffset = 0) {
        super(x, y, 0, scale, worldWidth, screenHeight, yOffset);
        this.name = name;
        this.baseWidth = 46;
        this.width = 46 * scale;

        // Emre Hoca faces left as requested and sits still
        this.facingLeft = true;
        this.state = 'idle';

        this.loadTextures();
    }

    async loadTextures() {
        if (!this.name) return;

        console.log(`👤 Loading TeacherNPC: ${this.name}...`);

        try {
            // Path: teachers/emre_hoca/idle/idle.png
            const idleImg = await Utils.loadImage(`teachers/${this.name}/idle/idle.png`);

            if (idleImg) {
                console.log(`✅ ${this.name} idle image loaded: ${idleImg.width}x${idleImg.height}`);
                this.baseWidth = idleImg.width;
                this.baseHeight = idleImg.height;
                this.width = this.baseWidth * this.scale;
                this.height = this.baseHeight * this.scale;

                this.animations.idle = [this.createCanvasFromImage(idleImg)];

                // Load walk sequence (0.png, 1.png, 2.png, 3.png)
                this.animations.walk = [];
                for (let i = 0; i < 4; i++) {
                    const walkFrame = await Utils.loadImage(`teachers/${this.name}/walk/${i}.png`);
                    if (walkFrame) {
                        this.animations.walk.push(this.createCanvasFromImage(walkFrame));
                    }
                }

                // Fallback for walk if sequence missing
                if (this.animations.walk.length === 0) {
                    this.animations.walk = [this.createCanvasFromImage(idleImg)];
                }

                this.loaded = true;
            } else {
                console.error(`❌ Teacher ${this.name} idle image NOT found at: teachers/${this.name}/idle/idle.png`);
                this.createFallbacks();
            }
        } catch (error) {
            console.error(`❌ Error loading textures for Teacher ${this.name}:`, error);
            this.createFallbacks();
        }
    }

    createCanvasFromImage(img) {
        const canvas = document.createElement('canvas');
        canvas.width = img.width;
        canvas.height = img.height;
        const ctx = canvas.getContext('2d');
        ctx.drawImage(img, 0, 0);
        return canvas;
    }

    update(dt) {
        // Teachers are static unless explicitly set to walk (for cinematics)
        if (this.state !== 'walk') {
            this.state = 'idle';
        } else {
            // If in walk state, update position based on speed/direction
            this.x += this.facingLeft ? -this.speed : this.speed;
        }

        // Update animation frame if multiple frames exist
        const anim = this.animations[this.state];
        if (anim && anim.length > 1) {
            this.frameIndex = (this.frameIndex + 0.1) % anim.length;
        }
    }
}

class MultiTeacherNPC extends NPC {
    constructor(x, y, index, scale = 4, worldWidth = 1536, screenHeight = 1024, yOffset = 0) {
        // Teachers: smaller scale (0.60x) and increased yOffset (+60) to fix floating/foot alignment
        const adjustedScale = scale * 0.60;
        const footFixOffset = yOffset + 60;

        super(x, y, index, adjustedScale, worldWidth, screenHeight, footFixOffset);

        // Force fully static behavior
        this.state = 'idle';
        this.laziness = 1.0;
        this.restlessness = 0.0;
    }

    // Override update completely to disable movement
    update(dt) {
        if (!this.loaded) return;

        // Force idle state
        this.state = 'idle';

        // Only update animation frame (if we had animation)
        const anim = this.animations.idle;
        if (anim && anim.length > 1) {
            this.frameIndex += (this.animationSpeed * 0.5) * (dt / 16.67);
            if (this.frameIndex >= anim.length) {
                this.frameIndex = 0;
            }
        }
    }

    async loadTextures() {
        try {
            const sheet = await Utils.loadImage('teachers/npcteachers/idle/0.png');

            if (sheet) {
                const totalTeachers = 6;
                const singleWidth = Math.floor(sheet.width / totalTeachers);
                const height = sheet.height;

                this.baseWidth = singleWidth;
                this.baseHeight = height;
                this.width = this.baseWidth * this.scale;
                this.height = this.baseHeight * this.scale;

                // Extract specific teacher frame using npcIndex
                const safeIndex = this.npcIndex % totalTeachers;
                const startX = Math.floor(safeIndex * singleWidth);

                const frame = this.extractFrame(sheet, startX, 0, singleWidth, height);

                this.animations.idle = [frame];
                this.animations.walk = [frame];

                this.loaded = true;
                console.log(`👨‍🏫 MultiTeacherNPC ${this.npcIndex} loaded.`);
            } else {
                console.warn('⚠️ Failed to load teacher sprite sheet (0.png).');
                this.createFallbacks();
            }
        } catch (error) {
            console.error('❌ Error loading MultiTeacherNPC:', error);
            this.createFallbacks();
        }
    }
}

File: raylib_port/src/npc.cpp
#include "npc.h"
#include <unordered_map>

static std::vector<Texture2D> cachedStudentIdleFrames[8];
static std::vector<Texture2D> cachedStudentWalkFrames[8];
static bool studentTexturesLoaded[8] = {false};

static std::unordered_map<std::string, std::vector<Texture2D>> cachedTeacherIdleFrames;
static std::unordered_map<std::string, std::vector<Texture2D>> cachedTeacherWalkFrames;
static std::unordered_map<std::string, bool> teacherTexturesLoaded;

NPC::NPC(float x, float y, int npcIndex, float scale, float worldW, float screenH, float yOffset)
    : x(x), y(y), npcIndex(npcIndex), scale(scale), worldWidth(worldW), screenHeight(screenH), yOffset(yOffset)
{
    baseWidth = 371.0f / 8.0f;
    baseHeight = 74.0f;
    width = 46.0f * scale;
    height = baseHeight * scale;
    state = "walk";
    facingLeft = GetRandomValue(0, 1) == 0;
    speed = 0.8f + Utils::RandomFloat(0, 0.8f);
    alive = true;
    moveOnlyLeft = false;
    wrapAround = false;
    frameIndex = Utils::RandomFloat(0, 4);
    animationSpeed = 0.12f;
    laziness = 0.002f + Utils::RandomFloat(0, 0.01f);
    restlessness = 0.005f + Utils::RandomFloat(0, 0.02f);
    loaded = false;
    leaving = false;
    LoadTextures();
}

NPC::~NPC() {
    // Textures are now managed globally and shared. Do not unload them here.
}

void NPC::LoadTextures() {
    int idx = npcIndex % 8;
    if (studentTexturesLoaded[idx]) {
        idleFrames = cachedStudentIdleFrames[idx];
        walkFrames = cachedStudentWalkFrames[idx];
        loaded = true;
        return;
    }

    if (FileExists("npc_textures/npc_idle.png") &&
        FileExists("npc_textures/npc_walk0.png") &&
        FileExists("npc_textures/npc_walk1.png") &&
        FileExists("npc_textures/npc_walk2.png"))
    {
        Texture2D idleSheet = LoadTexture("npc_textures/npc_idle.png");
        Texture2D w0 = LoadTexture("npc_textures/npc_walk0.png");
        Texture2D w1 = LoadTexture("npc_textures/npc_walk1.png");
        Texture2D w2 = LoadTexture("npc_textures/npc_walk2.png");

        int startX = (int)(idx * baseWidth);
        int frameW = (int)ceilf(baseWidth);

        if (idleSheet.id > 0) {
            idleFrames.push_back(Utils::ExtractFrame(idleSheet, startX, 0, frameW, (int)baseHeight));
            UnloadTexture(idleSheet);
        }
        if (w0.id > 0) {
            walkFrames.push_back(Utils::ExtractFrame(w0, startX, 0, frameW, (int)baseHeight));
            UnloadTexture(w0);
        }
        if (w1.id > 0) {
            walkFrames.push_back(Utils::ExtractFrame(w1, startX, 0, frameW, (int)baseHeight));
            UnloadTexture(w1);
        }
        if (w2.id > 0) {
            walkFrames.push_back(Utils::ExtractFrame(w2, startX, 0, frameW, (int)baseHeight));
            UnloadTexture(w2);
        }
        // Loop back
        if (walkFrames.size() >= 2) walkFrames.push_back(walkFrames[1]);
        
        cachedStudentIdleFrames[idx] = idleFrames;
        cachedStudentWalkFrames[idx] = walkFrames;
        studentTexturesLoaded[idx] = true;
        loaded = true;
    } else {
        // Fallback
        Color colors[] = {GREEN, BLUE, YELLOW, PURPLE, {50, 200, 200, 255}};
        Color c = colors[idx % 5];
        Image img = GenImageColor((int)baseWidth, (int)baseHeight, c);
        Texture2D tex = LoadTextureFromImage(img);
        UnloadImage(img);
        idleFrames.push_back(tex);
        walkFrames.push_back(tex);
        
        cachedStudentIdleFrames[idx] = idleFrames;
        cachedStudentWalkFrames[idx] = walkFrames;
        studentTexturesLoaded[idx] = true;
        loaded = true;
    }
}

void NPC::Update(float dt, float playerX, float playerY) {
    if (!loaded) return;

    if (state == "walk") {
        float dtScale = dt / 16.67f;
        x += (facingLeft ? -speed : speed) * dtScale;

        if (wrapAround) {
            if (x < -150) x = worldWidth + 150;
            else if (x > worldWidth + 150) x = -150;
        } else {
            if (x < 10) { x = 11; facingLeft = false; }
            else if (x > worldWidth - 10) { x = worldWidth - 11; facingLeft = true; }
        }

        if (!moveOnlyLeft && Utils::RandomFloat(0, 1) < laziness) state = "idle";
    } else {
        if (id.empty() && groupId.empty() && Utils::RandomFloat(0, 1) < restlessness) {
            state = "walk";
            if (!moveOnlyLeft && Utils::RandomFloat(0, 1) < 0.3f) facingLeft = !facingLeft;
        }
    }

    // Animate
    auto& frames = (state == "walk") ? walkFrames : idleFrames;
    frameIndex += animationSpeed * (dt / 16.67f);
    if (frameIndex >= frames.size()) frameIndex = 0;
}

void NPC::Draw(float dt) {
    if (!loaded || !alive) return;

    auto& frames = (state == "walk") ? walkFrames : idleFrames;
    if (frames.empty()) return;
    int idx = (int)frameIndex;
    if (idx >= (int)frames.size()) idx = 0;
    Texture2D& frame = frames[idx];

    float drawY = y + yOffset;
    Rectangle srcRect = {0, 0, (float)frame.width, (float)frame.height};
    if (facingLeft) srcRect.width = -(float)frame.width;
    Rectangle destRect = {x - width/2, drawY - height, width, height};
    DrawTexturePro(frame, srcRect, destRect, {0,0}, 0, WHITE);
}

Rect NPC::GetRect() const {
    return {x - width/2, y - height, width, height};
}

// ============ TeacherNPC ============
TeacherNPC::TeacherNPC(float x, float y, const std::string& name, float scale, float worldW, float screenH, float yOffset)
    : NPC(x, y, 0, scale, worldW, screenH, yOffset), name(name)
{
    facingLeft = true;
    state = "idle";
    idleFrames.clear();
    walkFrames.clear();
    LoadTextures();
}

void TeacherNPC::LoadTextures() {
    if (teacherTexturesLoaded[name]) {
        idleFrames = cachedTeacherIdleFrames[name];
        walkFrames = cachedTeacherWalkFrames[name];
        loaded = true;
        // set dimensions from cached frames
        if (!idleFrames.empty()) {
            baseWidth = (float)idleFrames[0].width;
            baseHeight = (float)idleFrames[0].height;
            width = baseWidth * scale;
            height = baseHeight * scale;
        }
        return;
    }

    std::string idlePath = TextFormat("teachers/%s/idle/idle.png", name.c_str());
    if (FileExists(idlePath.c_str())) {
        Texture2D tex = LoadTexture(idlePath.c_str());
        if (tex.id > 0) {
            baseWidth = (float)tex.width;
            baseHeight = (float)tex.height;
            width = baseWidth * scale;
            height = baseHeight * scale;
            idleFrames.push_back(tex);

            // Load walk frames
            for (int i = 0; i < 4; i++) {
                std::string wp = TextFormat("teachers/%s/walk/%d.png", name.c_str(), i);
                if (FileExists(wp.c_str())) {
                    walkFrames.push_back(LoadTexture(wp.c_str()));
                }
            }
            if (walkFrames.empty()) walkFrames = idleFrames;
            
            cachedTeacherIdleFrames[name] = idleFrames;
            cachedTeacherWalkFrames[name] = walkFrames;
            teacherTexturesLoaded[name] = true;
            
            loaded = true;
            return;
        }
    }
    // Fallback
    Image img = GenImageColor(46, 74, ORANGE);
    Texture2D tex = LoadTextureFromImage(img);
    UnloadImage(img);
    idleFrames.push_back(tex);
    walkFrames.push_back(tex);
    
    cachedTeacherIdleFrames[name] = idleFrames;
    cachedTeacherWalkFrames[name] = walkFrames;
    teacherTexturesLoaded[name] = true;
    
    loaded = true;
}

void TeacherNPC::Update(float dt, float playerX, float playerY) {
    if (state == "walk") {
        x += facingLeft ? -speed : speed;
    } else {
        state = "idle";
    }
    auto& frames = (state == "walk" && walkFrames.size() > 1) ? walkFrames : idleFrames;
    if (frames.size() > 1) {
        frameIndex = fmodf(frameIndex + 0.1f, (float)frames.size());
    }
}
